### PR TITLE
PerformContractMaintenanceRefactor

### PR DESCRIPTION
**Pull request by ChrisSchinnerl**
While doing some manual testing and debugging I noticed a few issues with our contract maintenance code that this PR is trying to improve.

- min score was calculated using only hosts that we don't have contracts with instead of all contracts
- contracts that were good but up for renewal immediately dropped out of the contract set upon failure to renew even though they shouldn't
- the logging in the contractor is lacking which makes it very hard to follow the contractor's reasoning
- the current structure of the contract maintenance code makes it very hard to write unit tests for it

Hopefully after this refactor we can get to a point where every piece of logic in the contractor can easily be unit tested as well as every decision followed by simply gepping for `performContractMaintenance` and the contract's ID.

### CODE REVIEW

File: autopilot/contractor/contractor.go [hunk 1]
```diff
@@ -189,714 +191,271 @@ func (c *Contractor) Close() error {
...
-
-	// no maintenance if no allowance was set
-	if cfg.Allowance.IsZero() {
-		return "allowance is set to zero, skipping contract maintenance", true
-	}
-
-	// no maintenance if no period was set
-	if cfg.Period == 0 {
-		return "period is set to zero, skipping contract maintenance", true
+func (c *Contractor) PerformContractMaintenance(ctx context.Context, w Worker, state *MaintenanceState) (bool, error) {
```
Comment (ChrisSchinnerl) [inline autopilot/contractor/contractor.go:194]: My recommendation when reviewing this is to just check it out locally and then start here. Everything that changed happens in here.

**Review by peterjan — COMMENTED** [review]
good stuff 👍🏻

File: autopilot/contractor/contractor.go [hunk 2]
```diff
@@ -1560,37 +889,535 @@ func renterFundsToExpectedStorage(renterFunds types.Currency, duration uint64, p
...
+	}
+
+	// STEP 3: perform contract formation
+	formedContracts, err := performContractFormations(ctx, bus, w, cr, ipFilter, logger, &remaining, int(ctx.WantedContracts())-len(keptContracts))
+	if err != nil {
+		return false, err
+	}
+
+	// fetch old set
+	oldSet, err := bus.Contracts(ctx, api.ContractsOpts{ContractSet: ctx.ContractSet()})
```
[the remark this answers is not in the document]
Comment (ChrisSchinnerl) [reply autopilot/contractor/contractor.go:1398]: This needs to happen before `SetContractSet` is called in step 4.

File: autopilot/contractor/ipfilter.go [hunk 3]
```diff
@@ -48,19 +41,11 @@ func (f *ipFilter) HasRedundantIP(host api.Host) bool {
...
-
 	return false
 }
 
-func (f *ipFilter) Remove(h api.Host) {
-	for k, v := range f.subnetToHostKey {
-		if v == h.PublicKey.String() {
-			delete(f.subnetToHostKey, k)
-		}
+func (f *ipFilter) Add(host api.Host) {
```
[the remark this answers is not in the document]
Comment (ChrisSchinnerl) [reply autopilot/contractor/ipfilter.go:47]: Changed it to `HostSet` but `Contains` seems a bit too misleading. Cause a host can be contained in the host set even though it's not actually contained.

File: autopilot/contractor/contractor.go [hunk 4]
```diff
@@ -1560,37 +889,535 @@ func renterFundsToExpectedStorage(renterFunds types.Currency, duration uint64, p
...
+	// fetch all hosts that are not blocked
+	hosts, err := bus.SearchHosts(ctx, api.SearchHostOptions{Limit: -1, FilterMode: api.HostFilterModeAllowed})
+	if err != nil {
+		return fmt.Errorf("failed to fetch all hosts: %w", err)
 	}
-	return time.Since(lastFailure) < failedRefreshForgivenessPeriod
+
+	var scoredHosts []scoredHost
+	for _, host := range hosts {
+		// filter out hosts that have never been scanned
```
[the remark this answers is not in the document]
Comment (ChrisSchinnerl) [reply autopilot/contractor/contractor.go:1263]: Probably not. I think it's a leftover from a while ago when we added the `scanned` field.

File: autopilot/contractor/contractor.go [hunk 5]
```diff
@@ -1560,37 +889,535 @@ func renterFundsToExpectedStorage(renterFunds types.Currency, duration uint64, p
...
+		// if the renewal/refresh failing was our fault (e.g. we ran out of
+		// funds), we should not drop the contract
+		if !usable && ourFault {
+			logger.Info("keeping contract even though renewal/refresh failed")
+			usable = true
+		}
+
+		// if the contract is not usable we ignore it
+		if !usable {
+			if inSet {
```
[the remark this answers is not in the document]
Comment (ChrisSchinnerl) [reply autopilot/contractor/contractor.go:1108]: They have different log levels. Theoretically we could still check and then assign a log level but this seemed more straightforward.

File: autopilot/contractor/contractor.go [hunk 6]
```diff
@@ -1560,37 +889,535 @@ func renterFundsToExpectedStorage(renterFunds types.Currency, duration uint64, p
...
 
-func (c *Contractor) HasAlert(ctx context.Context, id types.Hash256) bool {
-	ar, err := c.alerter.Alerts(ctx, alerts.AlertsOpts{Offset: 0, Limit: -1})
+// performContractChecks performs maintenance on existing contracts,
+// renewing/refreshing any that need it and filtering out contracts that should
+// no longer be used. The 'ipFilter' is updated to contain all hosts that we
+// keep contracts with and the 'dropOutReasons' map is updated with the reasons
+// for dropping out of the set. If a contract is refreshed or renewed, the
+// 'remainingFunds' are adjusted.
+func performContractChecks(ctx *mCtx, alerter alerts.Alerter, bus Bus, w Worker, cc contractChecker, cr contractReviser, ipFilter *ipFilter, dropOutReasons map[types.FileContractID]string, logger *zap.SugaredLogger, remainingFunds *types.Currency) ([]api.ContractMetadata, error) {
```
[the remark this answers is not in the document]
Comment (ChrisSchinnerl) [reply autopilot/contractor/contractor.go:898]: Theoretically, but right now the `Contractor` can be passed without modifications as a `contractReviser` and adding the `remainingFunds` to the `Contractor` isn't great. So we would need to introduce a wrapper type for the `Contractor` which has a `remainingFunds` field and then things get more verbose than just passing the pointer.

File: autopilot/contractor/contractor.go [hunk 7]
```diff
@@ -189,714 +191,271 @@ func (c *Contractor) Close() error {
...
-	// return whether the maintenance changed the contract set
-	return c.computeContractSetChanged(mCtx, currentSet, updatedSet, formed, refreshed, renewed, toStopUsing, contractData), nil
-}
-
-func (c *Contractor) computeContractSetChanged(ctx *mCtx, oldSet, newSet, formed []api.ContractMetadata, refreshed, renewed []renewal, toStopUsing map[types.FileContractID]string, contractData map[types.FileContractID]uint64) bool {
-	name := ctx.ContractSet()
+	// calculate the renter funds for the renewal a.k.a. the funds the renter will
+	// be able to spend
+	minRenterFunds, _ := initialContractFundingMinMax(ctx.AutopilotConfig())
+	renterFunds := renewFundingEstimate(minRenterFunds, contract.TotalCost, contract.RenterFunds(), log)
```
[the remark this answers is not in the document]
Comment (ChrisSchinnerl) [reply autopilot/contractor/contractor.go:399]: That is true but then unit testing becomes more annoying since we need to construct a full contract to check the 2 fields we are interested in. That's why I opted for minimal obsolete fields as a tradeoff on those lower level functions.

File: autopilot/contractor/contractor.go [hunk 8]
```diff
@@ -1560,37 +889,535 @@ func renterFundsToExpectedStorage(renterFunds types.Currency, duration uint64, p
...
+
+	// check if we want to run maintenance
+	if reason, skip := canSkipContractMaintenance(ctx, ctx.ContractsConfig()); skip {
+		logger.With("reason", reason).Info("skipping contract maintenance")
+		return false, nil
+	}
+	logger.Infow("performing contract maintenance")
+
+	// STEP 1: perform host maintenance
+	if err := performHostChecks(ctx, bus, logger); err != nil {
```
[the remark this answers is not in the document]
Comment (ChrisSchinnerl) [reply autopilot/contractor/contractor.go:1362]: We can do that but I'd make that a dedicated PR since we would probably move the pruning code out of the scanner and the host check code out of the contractor and into a new entity.

[the remark this answers is not in the document]
Comment (peterjan) [reply autopilot/contractor/contractor.go:1354]: Ok, problem with host checks is the (annoying) redundant IP field. That's also why currently `runHostChecks` is not persisting host checks until later that extra info gets annotated at which point we can persist them. Think you changed that though but it's still kind of annoying hard coupling between those two systems.

[the remark this answers is not in the document]
Comment (peterjan) [reply autopilot/contractor/contractor.go:1354]: I like the idea of having an extra `host` component though but we should probably rename the scanner and have it to everything, scan/check/prune. Right in time too because I _just_ refactored it!

File: autopilot/contractor/contractor.go (hunk 1 above)
Comment (peterjan) [reply autopilot/contractor/contractor.go:194]: I was going to open a PR for this today, but seeing as you have this huge `contractor` PR already perhaps you can tackle the following: we don't ever call `Close` on the contractor and the implementation of `Close` should probably also use a waitgroup to properly allow maintenance to finish gracefully.

**Review by peterjan — APPROVED** [review]
I approved it but perhaps we can tackle this comment here, if not I don't mind F/U with it. I was actually about to open a PR for that but then going through my list of reviews I found this one's still open.


https://github.com/SiaFoundation/renterd/pull/1367/files#r1696506803
